### ISSUE TITLE
Reducers on axis=0 need to be implemented and also fail more gracefully

### ISSUE THREAD

**Issue by iasonkrom** [dask-contrib/dask-awkward#256]
When events are read in `coffea` with `NanoEventsFactory.from_root()` with `permit_dask=True`, there are multiple operations with `dask-awkward` that should work in principle but actually do not.
I've made the following notebook to showcase these bugs because it would be too long to paste here:

https://nbviewer.org/urls/dl.dropbox.com/s/7fedcyb515u5qae/bugnofield.ipynb?flush_cache=true

The last cell of this notebook is also a feature request to support `axis=0` in `dak.count()`.

I've only had a brief chat with @lgray about this. In my understanding, these problems could also arise from minor bugs in `awkward` so I'm tagging  some people that might be interested to take a look:
@jpivarski @agoose77 @douglasdavis @martindurant

**Comment by lgray** [comment]
Here's a concise reproducer of the underlying issue:

both of the following dak reducer calls fail rather noisily:
```python3
import uproot
import dask_awkward as dak

events = uproot.dask({"https://github.com/CoffeaTeam/coffea/blob/master/tests/samples/nano_dy.root?raw=true": "Events"})

print(dak.num(events, axis=0).compute())

print(dak.count(events, axis=0).compute())
```

@iasonkrom Could you please make the title of this issue a bit more precise? The way it's put it seems like the sky is falling.

**Comment by lgray** [comment]
And moreover these functions, when I can coax them to work, seem to touch *all* the data in the object passed and they really shouldn't!

**Comment by lgray** [comment]
Heh - the failure is even more noisy in Iason's notebook :-)

**Comment by iasonkrom** [comment]
I would also like to point out the cells In [8] - In [11] in my notebook where this computation fails only if you call `dask.compute()` on the whole `nev` list but doesn't fail if you loop over the list or if you skip the first element.

**Comment by lgray** [comment]
Just to add a reference - we'd like to (rather heavily) use reducers on `axis=0` in https://github.com/CoffeaTeam/coffea/pull/797

**Comment by martindurant** [comment]
Is it the case, that reducers (let's say `sum`) with axis=0 can be implemented as a tree reduction:
```
ak.sum(arr, axis=0) => ak.sum(
    ak.concatenate(
        (ak.sum(part, axis=0) for part in array_partitions)
))
```
?

**Comment by agoose77** [comment]
@martindurant yes, in principle, although not without some extra massaging for the positional reducers like `argmax`, and handling `None`s.

**Comment by agoose77** [comment]
@martindurant [this Gist](https://gist.github.com/agoose77/595cc3e3d0c44059e2c6b1f1ad183c81) implements by-parts reduction, though _not_ as a tree reduction. The challenge is that positional reducers need auxillary information (the offsets) i.e. the result of `calculate_known_divisions`. That said, it should be fairly clear how this could be made into a tree reduction once `starts` are computed. Let me know if there is anything I can be doing here!

**Comment by martindurant** [comment]
I think it's reasonable for methods to require the division/partition offsets, and calculate them eagerly. There is a method for that. Often we know the values anyway, and we can clearly set expectations in the docs.

**Comment by agoose77** [comment]
The above reproducer:
https://github.com/dask-contrib/dask-awkward/issues/256#issuecomment-1536948630

Fails for two separate reasons that aren't inherently related to missing `axis=0` reducers.

1. `ak.count` cannot operate on ragged arrays if a non-negative axis is given. 
2. `ak.num` is not implemented as a reducer

We can fix `dak.num`, though, because it's using total reduction, which currently fails in certain scenarios. #267 could be the PR to address that, or another follow-up.

**Comment by iasonkrom** [comment]
Also here only the last line fails with
```
FieldNotFoundError: no field 'Electron' in record with 23 fields

This error occurred while attempting to slice

    <NanoEventsArray [<event 1:13889:3749778>, ...] type='40 * event'>

with

    'Electron'
```
```python
import awkward as ak
import dask_awkward as dak
import dask
from coffea.nanoevents import NanoAODSchema, NanoEventsFactory

fname = "coffea/tests/samples/nano_dy.root"
events = NanoEventsFactory.from_root(
    fname,
    schemaclass=NanoAODSchema,
    metadata={"dataset": "DYJets"},
    permit_dask=True,
).events()

twoelectron = dak.num(events.Electron) == 2
nomuon = dak.num(events.Muon) == 0

masks = [twoelectron, nomuon]

nev = [dak.sum(dak.from_awkward(ak.Array([40]), 1))]
nev.extend([dak.sum(mask) for mask in masks])

for i in nev:
    print(i.compute())

dask.compute(*nev[1:])
dask.compute(*nev)
```
while here
```python
import uproot
import awkward as ak
import dask_awkward as dak
import dask

events = uproot.dask({"coffea/tests/samples/nano_dy.root": "Events"})

twoelectron = dak.num(events.Electron_pt) == 2
nomuon = dak.num(events.Muon_pt) == 0

masks = [twoelectron, nomuon]
nev = [dak.sum(dak.from_awkward(ak.Array([40]), 1))]
nev.extend([dak.sum(mask) for mask in masks])

for i in nev:
    print(i.compute())

dask.compute(*nev[1:])
dask.compute(*nev)
```
the last line is the only one that works while the other 2 computes fail with
```
FieldNotFoundError: no field 'Electron_pt' in record with 1 fields

This error occurred while attempting to slice

    <Array [{run: 1}, {run: 1}, ..., {...}, {run: 1}] type='40 * {run: uint32}'>

with

    'Electron_pt'
```

**Comment by lgray** [comment]
@iasonkrom if you can try to remove coffea from the reproducers, it would be super helpful! If you can show it's just a problem with uproot/dask-awkward/awkward then the solutions are much easier to find. Nanoevents is complex enough to make ruling out problems with it, as opposed to dependencies, is hard to do on first glance. 

@agoose77 Any possibility to fix the typetracer warnings? Usually they come with some odd performance penalties. 

The latest reproducer seems a lot like a "form rehydration" issue, also encountered over in repartition/slicing.

**Comment by iasonkrom** [comment]
@lgray some of the snippets are only `uproot`. I was just trying to recreate similar things with and without `coffea`. The last snippet is super weird to me because only `dask.compute(*nev)` works but the for loop and `dask.compute(*nev[1:])` don't.
If the elements of the `nev` list are computed individually with the for loop, or if `dak.sum(dak.from_awkward(ak.Array([40]), 1))` is  not included in the list that you call `dask.compute()` on, it fails. In the for loop it fails after the first step of the loop so only `40` is printed out. The whole `nev` list is computed fine though.

Moreoever, computing the masks themselves fail.

These fail:
```python
import uproot
import dask_awkward as dak

events = uproot.dask({"coffea/tests/samples/nano_dy.root": "Events"})
twoelectron = dak.num(events.Electron_pt) == 2
twoelectron.compute()
```
```python
import uproot
import awkward as ak
import dask_awkward as dak
import dask

events = uproot.dask({"coffea/tests/samples/nano_dy.root": "Events"})
twoelectron = dak.num(events.Electron_pt) == 2
nev = [dak.sum(dak.from_awkward(ak.Array([40]), 1))]
nev.append(dak.sum(twoelectron))
print(nev[0].compute()) # this doesn't fail
print(nev[1].compute()) # fails here
```
This doesnt:
```python
import uproot
import awkward as ak
import dask_awkward as dak
import dask

events = uproot.dask({"coffea/tests/samples/nano_dy.root": "Events"})
twoelectron = dak.num(events.Electron_pt) == 2
nev = [dak.sum(dak.from_awkward(ak.Array([40]), 1))]
nev.append(dak.sum(twoelectron))
dask.compute(*nev)
```

**Comment by iasonkrom** [comment]
Following the latest comment on #267 by @chrispap95, I report that the behavior of my last comment here (https://github.com/dask-contrib/dask-awkward/issues/256#issuecomment-1557629209) is exactly the same.
I may be mistaken but could this be not a problem of the reducers, but a problem of how `uproot.dask()` touches the data?
@lgray @agoose77

**Comment by lgray** [comment]
uproot.dask is not responsible for touching data! It's all in dask-awkward/awkward!

**Comment by lgray** [comment]
While it's good for testing, my opinion is we cannot accept a solution until column-opt works in this situation.

**Comment by agoose77** [comment]
These are separate issues, in my view. Right now, column optimisation is somewhat "hazmat" - it isn't designed to work in scenarios that the columns are needed at a metadata level. #184 understands the distinction, and introduces empty

**Comment by iasonkrom** [comment]
Btw, disabling column optimization fixes all 3 snippets in https://github.com/dask-contrib/dask-awkward/issues/256#issuecomment-1557629209, but it doesn't fix anything when `coffea` is involved.
For instance the following snippet fails when `awkward.optimization.enabled` is set to `False` but works fine when it is set to `True`.

```python
import uproot
import awkward as ak
import dask_awkward as dak
import dask

from coffea.nanoevents import NanoAODSchema, NanoEventsFactory

with dask.config.set({"awkward.optimization.enabled": False}): 
    fname = "coffea/tests/samples/nano_dy.root"
    events = NanoEventsFactory.from_root(
        fname,
        schemaclass=NanoAODSchema,
        metadata={"dataset": "DYJets"},
        permit_dask=True,
    ).events()

    
    twoelectron = dak.num(events.Electron) == 2
    print(twoelectron.compute())
```

I don't know if I'm helping at all. I'm just trying to give working and failing reproducers for you guys whilst testing when new commits are being pushed in #267.

**Comment by agoose77** [comment]
I think this can be closed now?

**Comment by iasonkrom** [comment]
All the failures mentioned here were fixed in #267 and #184. Closed.
